### ISSUE TITLE
Implement ternary conditional expressions in front-end

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#613]
This is purely a front-end feature.

Should probably be done after #101.

Difficult part is dealing with parsing ambiguity between `:` as else in conditional and `:` between prefix and identifier.

**Comment by jclark** [comment]
Idea for how to handle ambiguity:

- Functions parsing expressions get an additional argument that is a optional record with two integer fields, elseColonIndex and colonCount
- This argument is non-nil in places where `:` could be the else in a conditional expr
- When parsing a colon that might be a prefix of an identifier, check whether elseColonIndex == colonCount; if they are equal, don't parse the colon as part of a prefix and return back to caller; otherwise treat the colon as part of the prefix as normal; in either case increment colonCount
- When parsing a conditional expression, after parsing the `?`, set i to 0 and repeatedly do the following
    - try parsing thenExpr `:` elseExpr
    - thenExpr is passed a record with elseColonIndex = i and colonCount = 0
    - this record is not passed to subexpressions within brackets
    - if parsing succeeds for both thenExpr and elseExpr and the next token after elseExpr is something that it is grammatically possible (e.g. `)` or `;` but not another binary operator), then return with this as the parse
    - otherwise if colonCount > elseColonIndex after parsing thenExpr, increment i and repeat
 

This is relying on our machinery to save and restore tokenizer state.

**Comment by jclark** [comment]
Maybe better to add a ParseContext argument to parsing functions and put the colon-related fields in that.

**Comment by jclark** [comment]
For example, consider

```
int v = x ? a:b : c:d ;
```

First try we will parse `a` as thenExpr, `b : c` as elseExpr. We will end up with the next token as `:` which will cause us to retry. Second try we will parse `a:b` as thenExpr, `c:d` as elseExpr, and the next token will the `;`, which is OK and will cause us to return.

**Comment by jclark** [comment]
@lochana-chathura Would https://github.com/ballerina-platform/nballerina/issues/613#issuecomment-949123547 work for jBallerina?

**Comment by jclark** [comment]
@hasithaa mentioned this example  `x ? a:b:c` which the above algorithm would parse as ` x ? (a) : (b:c)` but jBallerina parses as `x ? (a:b) : c`. I believe both parses are consistent with `m:x` having higher precedence than `x?y:z` since they both produce a parse where the outer expression is `x?y:z` and the inner expression is `m:x`.

**Comment by jclark** [comment]
Here's a more complicated one:

```
int v = x ? y + a:b + c + d + 2:c;
```

There's no ambiguity here, so the right parse is:

```
int v = x ? (y + a:b + c + d + 2):c;
```

which I believe the above algorithm delivers.

**Comment by jclark** [comment]
The spec says that `x?y:z`  is right associative which means `a ? b : c ? d : e` parses as `a ? b : (c ? d : e)` rather than `(a ? b : c) ? d : e`.  To me it feels more consistent with that to treat `a ? b : c : d` as `a ? b : (c : d)` than `a ? (b : c) : d`.

**Comment by lochana-chathura** [comment]
> @lochana-chathura Would [#613 (comment)](https://github.com/ballerina-platform/nballerina/issues/613#issuecomment-949123547) work for jBallerina?

The aforementioned algorithm does not give the same result as our current jballerina implementation. 

What we do currently is, equivalent to this,
- When parsing a conditional expression, after parsing the `?`, 
    - try parsing thenExpr (meanwhile keep the colonCount)
    - colonCount is not passed to subexpressions within brackets.
    - see if the next token is a colon. if yes, return with the parse.
    - if no, then try parsing again and stop the thenExpr parsing at the colonCount and return the parse.

e.g. 
1. `x ? a:b:c;`
if we try parsing thenExpr it would be `a:b` and next token would be `:`. so it is parsed as `(a:b) : c`
similarly `x ? a:b : c:d ` would parse as `x ? (a:b) : (c:d)`

2. `x ? y + a:b + c + d + z:c;`
if we try parsing thenExpr it would be `y + a:b + c + d + z:c` and next token is not a colon `:`. So in the next time we stop thenExpr parsing at the colonCount(second colon). So we get `x ? (y + a:b + c + d + z):c`

And I think the above parsing is consistent with `m:x` having higher precedence over the `x?y:z`.

**Comment by jclark** [comment]
> if we try parsing thenExpr it would be `y + a:b + c + d + 2:c`

How do you get that? That's not a syntactically valid expression. Why wouldn't you get `y + a:b + c + d + 2` immediately?

**Comment by lochana-chathura** [comment]
> > if we try parsing thenExpr it would be `y + a:b + c + d + 2:c`
> 
> How do you get that? That's not a syntactically valid expression. Why wouldn't you get `y + a:b + c + d + 2` immediately?

sorry, my bad. In `y + a:b + c + d + 2:c`, thenExpr would be `y + a:b + c + d + 2` and resolves similar to sample 1.

the example I wanted to take is `y + a:b + c + d + z:c`.
In that thenExpr would be `y + a:b + c + d + z:c`

**Comment by jclark** [comment]
The algorithms are pretty similar. The difference is:

a) jBallerina algorithm is parsing as much as possible and then, if that doesn't work, parsing less 
b) my algorithm is parsing as little as possible and then, if that doesn't work, parsing more

I suspect with both algorithms there are cases where you need to retry more than once...

**Comment by lochana-chathura** [comment]
> The algorithms are pretty similar.
> I suspect with both algorithms there are cases where you need to retry more than once...

Agreed. In our implementation, we do the parse only once and if that doesn't work, we split the parsed node at the rightmost qualified identifier and rearrange the node. I guess it is not possible with nballerina parser.

**Comment by jclark** [comment]
How does the jBallerina algorithm apply to  `int v = a ? b : c ? d : e;`?

**Comment by lochana-chathura** [comment]
> How does the jBallerina algorithm apply to `int v = a ? b : c ? d : e;`?

With right associativity, It would first parse as `a ? [(b : c) ? d : e]`. Here the thenExpr would be `(b : c) ? d : e`. 

Then we split the node at the right-most qualified-identifier. In this example, we only have one qualified-identifier within thenExpr which is `b:c`. So we split there. The result would be `a ? b : (c ? d : e)`.

I think we will have an equivalent behaviour, if the `colonCount` is counted only for qualified identifiers when trying to parse thenExpr. So in the second parse, we will return after b in the above example.

**Comment by jclark** [comment]
See https://github.com/ballerina-platform/ballerina-spec/issues/1004

#734 helps but does not solve the problem
